multisig-wallet/src/main.sw

```diff
@@ -0,0 +1,141 @@
+contract;
+
+use std::{
+    address::Address,
+    assert::require,
+    b512::B512,
+    context::call_frames::contract_id,
+    contract_id::ContractId,
+    ecr::{EcRecoverError, ec_recover_address},
+    hash::sha256,
+    logging::log,
+    result::*,
+    revert::revert,
+    storage::{get, store}
+};
+
+use core::num::*;
+
+abi MultiSignatureWallet {
+    fn constructor(owner1: Address, owner2: Address) -> bool;
+    fn execute_transaction(to: ContractId, value: u64, data: b256, signature1: B512, signature2: B512) -> bool;
+    fn is_owner(owner: Address) -> bool;
+    fn get_transaction_hash(to: ContractId, value: u64, data: b256, nonce: u64) -> b256;
+}
+
+enum Error {
+    CannotReinitialize: (),
+    IncorrectSignerOrdering: (),
+    NotAnOwner: (),
+    NotInitialized: (),
+}
+
+struct Tx {
+    destination: b256,
+    value: u64,
+    data: b256,
+    nonce: u64,
+    contact_identifier: b256,
+}
+
+storage {
+    /// Used to add entropy into hashing of Tx to decrease the probability of collisions
+    nonce: u64,
+
+    /// The number of approvals (signatures * weight) required in order to execture a Tx
+    threshold: u64,
+}
+
+impl MultiSignatureWallet for Contract {
+    /// Initializes the contract by setting the owners, owner weightings and initializes the nonce
+    ///
+    /// # Panics
+    ///
+    /// - When the constructor is called more than once
+    fn constructor(owner1: Address, owner2: Address) -> bool {
+        require(storage.nonce == 0, Error::CannotReinitialize);
+
+        // TODO: when vectors are implemented change owners to be a Vec<Address>
+        store(owner1.value, true);
+        store(owner2.value, true);
+
+        // TODO: Weighted signers https://github.com/FuelLabs/sway-applications/issues/4
+
+        storage.nonce = 1;
+        true
+    }
+
+    /// Executes a Tx if the required signatures meet the restrictions on ownership and threshold approval
+    ///
+    /// # Panics
+    ///
+    /// - When the constructor has not been called to initialize the contract
+    /// - When the public key cannot be recovered from a signature
+    /// - When the signer is not an owner
+    /// - When the signatures are not in ascending order (0x1 < 0x2 < 0x3...)
```

Comment:
I see how the comparison is done below, which must mean Ord is implemented for `B512`.

But what would force `signer1.value < signer2.value` to be the case?

Reply:
Ok. Same question then I guess, how are the addresses being in ascending order enforced when an address is just a hash of a pub key derived from randomness ultimately?

Reply:
> I see how the comparison is done below, which must mean Ord is implemented for `B512`.
> 
> But what would force `signer1.value < signer2.value` to be the case? I'm sure it's here, just not seeing it at first pass. The incrementing nonce must be a part of this, but the signature is otherwise not sequential I would think.

My reference implementation was to follow [this execute function](https://github.com/SilentCicero/MultiSignatureWallet#reference-implementation-solidity)

Reply:
You can enforce they're in ascending order because they're public.

Reply:
The signatures would need to be placed in the function in such an order that the recovered address values are ascending in order otherwise it would error out because of a possible single owner trying to execute the Tx with their own signature multiple times

Reply:
Oh, if there are two sigs then indeed an inequality is sufficient. But eventually this will support an arbitrary number of signers.